### PR TITLE
Topic recognition - ADNI GFnet (ViT) Transformer (s47018457)

### PR DESCRIPTION
**Pull request by LachlanMcKay**
Hello Teaching team 

The task I chose was to make a GFnet (Vision Transformer) to classify Alzheimers Disease (AD) or normal (NC) in the dataset. I manage to get an accuracy of 66.7% in my test split but in my training / validation I got approximately  96.50%. The files included are as follows 

images/: Contains the imaged used in my README.md file

module.py: This file contains the structure of my Vision transformer as well as the data augmentation I did on the data

dataset.py: This file contains data loaders and obtaining the datasets pathway

train.py: This file houses the training script and how to evaluate the model after training. 

predict.py: This file contains the hyper-parameters of the model, triggers the training and evaluation stages, and saves plots for the obtained results

This task was done in PyTorch.

Thank You. 
Lachlan McKay s47018457

### CODE REVIEW

**Comment by wangzhaomxy** [comment]
**<This is an initial inspection, no action is required at this point.>**

**File Organizing:** Wrong file path. All student-constructed documents should be packed in the ‘PatternAnalysis-2024/recognition/<student_constructed_folder>/’ folder. Please reorganize your files, otherwise it won't be merged into Shakes' branch. **<Student should be acting in this part>**

**Problem Solving:**

- The algorithm can solve the problem.
- Accuracy in the testing dataset is claimed to be 0.64, but no evidence supports this result. 

**Model and functions:**

- It correctly uses PyTorch to construct the GFNet models. 
- With data augmentation.

**Code design:** Poor.

- data preprocessing functions, such as transforms, should be documented in the dataset.py instead of modules.py.
- The executable training code is not organized in the train.py but predict.py.
- The helper functions are organized in the modules.py instead of util.py.
- The executable code should be involved in the predict.py to predict labels from the input image.

**Code comment and docstring:**

- Good code comments 
- Multiple styles of function docstrings

**Difficulty:** Hard.

**Additional Comments:**

- Insufficient commits, just copy-and-paste code.
- Poor ReadMe design: it lacks visual aids and the writing style is not suitable for a formal report.
- Reference from website should contain a title instead of just a link.

**Comment by aniketgupta17** [comment]
Observational Feedback

Pull Request:
Correctly created the Pull request from Topic Recognition Branch .
The pull request includes a clear description about the file structure .
**No previous feedback incorporated .** 

File Organizing: Wrong file path. All student-constructed documents should be packed in the ‘PatternAnalysis-2024/recognition/<student_constructed_folder>/’ folder. Please reorganize your files, otherwise it won't be merged into Shakes' branch.

Commit Log:
Commit messages could be more progressive for the Recognition Problem solved using 4 files . 
Commits are not regularly made.

Documentation:
The README file **should** contain in depth  explanation of the algorithm and the problem it solves.
Code comments and docstrings  are included.
Proper GitHub markdown formatting is not used, without  organized headings, lists, and code blocks.
Can add about Conclusion and  future improvements in the ReadMe.

**Comment by LachlanMcKay** [comment]
Am I able to implement the feedback . Notably these comments 

"Poor ReadMe design: it lacks visual aids and the writing style is not suitable for a formal report.
Reference from website should contain a title instead of just a link." - wangzhaomxy

"Proper GitHub markdown formatting is not used, without organized headings, lists, and code blocks.
Can add about Conclusion and future improvements in the ReadMe." -  aniketgupta17

Also I wasn't able to get my matplotlib plots to get saved. I do have the .out file with the results should this be included?

Thank you for your time

**Comment by wangzhaomxy** [comment]
Sorry. Any modification to the code and report is not accepted now. Shakes will give you marks and tell you to change some merge errors if there are any.
Cheers,
Zhao

**Comment by gayanku** [comment]
## Marking
<table><tbody>
  <tr><td colspan='3'><b>Good/OK/Fair Practice (Design/Commenting, TF/Torch Usage)</b></td> </tr>
	<tr><td></td><td>Adequate design and implementation. </td><td>-1</td></tr>
	<tr><td></td><td>Spacing and comments. </td><td></td></tr>
	<tr><td></td><td>No Header blocks. </td><td>-1</td></tr>
	
  <tr><td colspan='3'><b>Recognition Problem</b></td></tr>
	<tr><td></td><td>OK solution to problem. Claimed acc not supported in report.</td><td>-1</td></tr>
	<tr><td></td><td>Driver Script present. </td><td></td></tr>
	<tr><td></td><td>File structure NOT present. </td><td>-1</td></tr>
	<tr><td></td><td>Good Usage & Demo & Visualisation & Data usage. </td><td></td></tr>
	<tr><td></td><td>Module present. </td><td></td></tr>
	<tr><td></td><td>Commenting present. </td><td></td></tr>
	<tr><td></td><td>No Data leakage found. Uses Data Aug.</td><td></td></tr>
	<tr><td></td><td>Difficulty : Hard.  GFNet (Hard Difficulty)</td><td></td></tr>
	
  <tr><td colspan='3'><b>Commit Log</b></td></tr>
	<tr><td></td><td>Some/Adequate Meaningful commit messages. </td><td>-1</td></tr>
	<tr><td></td><td>Good Progressive commits. </td><td></td></tr>
	
  <tr><td colspan='3'><b>Documentation</b></td></tr>
	<tr><td></td><td>Readme :Acceptable. </td><td>-2</td></tr>
	<tr><td></td><td>Model/technical explanation :Acceptable. </td><td>-1</td></tr>
	<tr><td></td><td>Description and Comments :Acceptable. Proper references needed.</td><td>-1</td></tr>
	<tr><td></td><td>Markdown used and PDF submitted. PDF Checked.</td><td></td></tr>
	
  <tr><td colspan='3'><b>Pull Request</b></td></tr>
	<tr><td></td><td>Pull Request has problems: Incorrect branch.</td><td>-1</td></tr>
	<tr><td></td><td>Feedback action required: Feedback marks possible +2 if the requested changes are made. Needs stucture changed and branch fixed.</td><td>-2</td></tr>
	<tr><td></td><td>Request Description is good. </td><td></td></tr>
  <tr><td colspan='2'><b>TOTAL</b></td><td>-12</td></tr>
</tbody></table>

Marked as per the due date and changes after which aren't necessarily allowed to contribute to grade for fairness.
_Subject to approval from Shakes_

**Comment by gayanku** [comment]
Feedback marks possible +2 if the requested changes are made (see above).

**Comment by LachlanMcKay** [comment]
I believe I have incorporated the feedback which was to fix the file structure and branch. Please let me know if there is anything I need to do or have done so incorrectly

**Comment by shakes76** [comment]
Feedback completed +2, no plots and other visulisations -2.

**Comment by shakes76** [comment]
Cant merge because of conflicting changes to main repo files (README). Please update for merge, doesn't affect grade.
